README.md
# 3Box Address Server

[![Greenkeeper badge](https://badges.greenkeeper.io/3box/3box-address-server.svg)](https://greenkeeper.io/)

`3box-address-server` is a server utilizing a REST-API that is used to associate an Ethereum address with its root OrbitDB address. This is what must be looked up to sync the user's data.

## API Endpoint

```
https://beta.3box.io/address-server/
```

## API Description


### Set

`POST /odbAddress`

#### Body

```
{
    address_token: <jwt token containing the orbitDB root store address>
}
```

The `address_token` is a [DID signed jwt](https://github.com/uport-project/did-jwt.git) of the address to be published. The payload of the `address_token` is:
```
{
    rootStoreAddress: <orbitdb root store address>
}
```

#### Response

| Status |     Message    |                                                   |
|:------:|----------------|---------------------------------------------------|
| 200    | Ok.            | rootStoreAddress stored                           |
| 401    | Invalid JWT    | Posted token is invalid (signature, expired, etc) |
| 403    | Missing data   | no `rootStoreAddress` in `address_token`          |
| 500    | Internal Error | Internal Error                                    |

The response data follows the [`jsend`](https://labs.omniti.com/labs/jsend) standard.

#### Response data
```
{
  status: 'success',
  data: {
    rootStoreAddress: <the orbitDB root store address that was accepted>
  }
}
```

### Link an ethereum address to a DID

`POST /link`


#### Body

```
{
    consent_signature: <signature>,
    linked_did: <DID>,
    consent_msg: <string-message>,
    type: <string-[ethereum-eoa, erc1271]>,
    chainId: <int>,
    address: <string-smart account address>,
    timestamp: <int-unix timestamp>
}
```

The `consent_signature` is a [personal_sign signature](https://web3js.readthedocs.io/en/1.0/web3-eth-personal.html) of a consent message and the DID to be linked. The data(text) of the signature is:

```
Create a new 3Box profile

-
Your unique profile ID is did:muport:Qmsd89f7hg0w845hsdd
```


The ethereum address to be linked is recovered from the signature.

#### Response

| Status |     Message     |                                                  |
|:------:|-----------------|--------------------------------------------------|
| 200    | Ok.             | Link created and stored                          |
| 400    | Bad request     | No did on the message or dids does not match     |
| 401    | Invalid consent | Posted signature is invalid (wrong DID, etc)     |
| 403    | Missing data    | no `consent_signature` or `linked_did`           |
| 500    | Internal Error  | Internal Error                                   |

The response data follows the [`jsend`](https://labs.omniti.com/labs/jsend) standard.

#### Response data
```
{
  status: 'success',
  data: {
    did: <the did that was linked>,
    address: <the ethereum address that was linked>
  }
}
```

### Delete a link between an ethereum address and a DID

`POST /linkdelete`

#### Body

```
{
  delete_token: <jwt token>
}
```

The `delete_token` is a [DID signed jwt](https://github.com/uport-project/did-jwt.git) with a 1 hour expiry, of the address to be deleted. The payload of the `delete_token` is:

```
{
  "address": "0x19bd82476fa8799e0eb7bbb03ee2a67678c01cdc",
  "type": "delete-address-link"
}
```

#### Response

| Status |     Message     |                                                  |
|:------:|-----------------|--------------------------------------------------|
| 200    | Ok.             | Link deleted                                     |
| 401    | Invalid consent | delete token failed verification                 |
| 403    | Missing data    | no delete_token                                  |
| 500    | Internal Error  | Internal Error                                   |

The response data follows the [`jsend`](https://labs.omniti.com/labs/jsend) standard.

#### Response data
```
{
  status: 'success',
  data: 'address of link deleted'
}
```

### Get orbitDB root store address for given identity

`GET /odbAddress/{identity}`

Here the `ìdentity` is either a `DID` or an `ethereum address`.

#### Response

| Status |     Message     |                                                  |
|:------:|-----------------|--------------------------------------------------|
| 200    | Ok.             | An orbitDB address is returned                   |
| 404    | Not found       | The DID or ethereum address was not found        |
| 500    | Internal Error  | Internal Error                                   |

The response data follows the [`jsend`](https://labs.omniti.com/labs/jsend) standard.

#### Response data
```
{
  status: 'success',
  data: {
    rootStoreAddress: <the orbitDB root store address associated with the identity>
  }
}
```


### Get multiple orbitDB root store addresses for given identities

`Post /odbAddresses`

Returns multiple root store addresses from multiple identities. If some identities are present but not all, the response will only contain the ones that do have root store addresses associated with them.

#### Body

```
{
    identities: <an array containing identities as strings>
}
```

Here the `ìdentity` is either a `DID` or an `ethereum address`.

#### Response

| Status |     Message     |                                                    |
|:------:|-----------------|----------------------------------------------------|
| 200    | Ok.             | All or some of the orbitDB addresses are returned  |
| 400    | Bad request     | No identities found in request                     |
| 404    | Not found       | No matching orbitDB addresses where found          |
| 500    | Internal Error  | Internal Error                                     |

The response data follows the [`jsend`](https://labs.omniti.com/labs/jsend) standard.

#### Response data
```
{
  status: 'success',
  data: {
    rootStoreAddresses: {
      <identity>: <the orbitDB root store address associated with the identity, or null if no match for this identity>
    }
  }
}
```

## Maintainers
[@simonovic86](https://github.com/simonovic86)


CHANGELOG.md
## v1.4.0 - 2020-07-09
* feat: upgrade s3 ipfs repo, shard blockstore support

## v1.3.0 - 2020-05-01
* chore: upgrade did-resolver and did-jwt libraries

## v1.2.0 - 2020-03-12
* feat: add bunyan logger
* feat: add fetch mutliple links and root store address queries

## v1.1.6 - 2020-01-24
* fix: don't verify EOA type signature for erc1271 contracts

## v1.1.5 - 2020-01-15
* fix: let AWS S3 client load credentials from environment variables

## v1.1.4 - 2020-01-07
* feat: load secrets from env vars only (if valid)
* feat: add dockerfile for local development using serverless offline
* feat: allow S3 client options to be set for endpoint, addressing style and signature version

## v1.1.3 - 2019-12-16
* feat: configure separate dev and prod environments
* feat: add parameter validation for ethereum addresses
* feat: re-enable CI
* security: upgrade webpack to v4.41.3


src/api/link_post.js
const isIPFS = require('is-ipfs')
const ethers = require('ethers')

const ACCOUNT_TYPES = {
  ethereum: 'ethereum',
  ethereumEOA: 'ethereum-eoa',
  erc1271: 'erc1271',
  conflux: 'conflux'
}

const SUPPORTED_CHAINS = {
  1: 'homestead',
  3: 'ropsten',
  4: 'rinkeby',
  42: 'kovan',
  5: 'goerli'
}

const MAGIC_ERC1271_VALUE = '0x20c13b0b'

// valid 3 DID or muport DID
const isValidDID = did => {
  const parts = did.split(':')
  if (!parts[0] === 'did' || !(parts[1] === '3' || parts[1] === 'muport')) return false
  return isIPFS.cid(parts[2])
}

const isValidAccountType = type => {
  return Object.values(ACCOUNT_TYPES).includes(type)
}

const isSupportedChainId = chainId => {
  return chainId && Boolean(SUPPORTED_CHAINS[chainId])
}

const isValidTimestamp = timestamp => {
  return parseInt(timestamp, 10) > 0 && parseInt(timestamp, 10) <= +new Date()
}

const isValidSignature = async (contractAddress, chainId, sig, msg) => {
  const abi = [
    'function isValidSignature(bytes _messageHash, bytes _signature) public view returns (bytes4 magicValue)'
  ]
  const network = SUPPORTED_CHAINS[chainId]
  const ethersProvider = ethers.getDefaultProvider(network)
  const contract = new ethers.Contract(contractAddress, abi, ethersProvider)
  const message = '0x' + Buffer.from(msg, 'utf8').toString('hex')
  const returnValue = await contract.isValidSignature(message, sig)

  return returnValue === MAGIC_ERC1271_VALUE
}

class LinkPostHandler {
  constructor (sigMgr, linkMgr) {
    this.sigMgr = sigMgr
    this.linkMgr = linkMgr
  }
  async handle (event, context, cb) {
    // Parse body
    let body
    try {
      body = JSON.parse(event.body)
    } catch (e) {
      cb({ code: 403, message: 'no json body: ' + e.toString() })
      return
    }

    const handlers = {
      '0':  this.v0Handler,
      '1':  this.v1Handler,
    }

    const version = body.version || 0

    if (!Object.keys(handlers).includes(version.toString())) {
      cb({ code: 403, message: 'invalid link proof version' })
      return
    }

    let {
      msg,
      sig,
      did,
      type,
      chainId,
      contractAddress,
      timestamp,
    } = await handlers[version](body, cb)

    // Get address from signature + msg
    const address = type === ACCOUNT_TYPES.erc1271 ? contractAddress : await this.sigMgr.verify(msg, sig)
    const consent = JSON.stringify({ msg, sig })

    await this.linkMgr.store(address, did, consent, type, chainId, contractAddress, timestamp)

    cb(null, { did: did, address: address })
  }

  async v0Handler(body, cb) {
    // Check if data is present
    if (!body.consent_signature) {
      cb({ code: 403, message: 'no consent_signature' })
      return
    }
    if (!body.consent_msg) {
      cb({ code: 403, message: 'no consent_msg' })
      return
    }
    if (!body.linked_did) {
      cb({ code: 403, message: 'no linked_did' })
      return
    }

    const sig = body.consent_signature
    const msg = body.consent_msg
    const did = body.linked_did

    const regex = /Your unique profile ID is (.*)/

    let msg_did = regex.exec(msg)
    if (!msg_did) {
      cb({ code: 400, message: 'no did on the consent_msg' })
      return
    }

    if (msg_did[1] !== did) {
      cb({ code: 400, message: 'dids does not match' })
      return
    }

    return {
      msg,
      sig,
      did,
      type: ACCOUNT_TYPES.ethereumEOA,
      chainId: null,
      contractAddress: null,
      timestamp: null
    }
  }

  async v1Handler(body, cb) {
    // Check if data is present
    if (!body.signature) {
      cb({ code: 403, message: 'no signature' })
      return
    }
    if (!body.message) {
      cb({ code: 403, message: 'no message' })
      return
    }

    const sig = body.signature
    const msg = body.message
    const type = body.type
    const chainId = body.chainId
    const address = body.address
    const timestamp = body.timestamp

    const regex = /(did:\S+)/

    const didMatch = regex.exec(msg)
    if (!didMatch || !isValidDID(didMatch[0])) {
      cb({ code: 400, message: 'no valid did in the consent_msg' })
      return
    }
    const did = didMatch[0]

    if (Boolean(chainId) && !isSupportedChainId(chainId)) {
      cb({ code: 400, message: 'unsupported chainId provided' })
      return
    }

    if (Boolean(type) && !isValidAccountType(type)) {
      cb({ code: 400, message: 'unsupported type provided' })
      return
    }

    if (type === ACCOUNT_TYPES.erc1271 && !address) {
      cb({ code: 403, message: 'Parameter "address" needed for type erc1271' })
      return
    }

    if (type === ACCOUNT_TYPES.erc1271 && !(await isValidSignature(address, chainId, sig, msg))) {
      cb({ code: 400, message: 'invalid signature provided' })
      return
    }

    if (Boolean(timestamp) && !isValidTimestamp(timestamp)) {
      cb({ code: 400, message: 'invalid timestamp provided' })
      return
    }

    return {
      msg,
      sig,
      did,
      type: type || ACCOUNT_TYPES.ethereumEOA,
      chainId: chainId || null,
      contractAddress: address || null,
      timestamp: timestamp || null
    }
  }
}
module.exports = LinkPostHandler
